File: komodo/dag.py
"""Dependency ordering and wave scheduling over a group's open tasks."""

from __future__ import annotations

from typing import Dict, List, Sequence

from .tasks import Task


class CycleError(ValueError):
    """Raised when depends_on edges form a loop."""


def topological(tasks: Sequence[Task]) -> List[Task]:
    """Tasks ordered so every dependency precedes its dependents, file order breaking ties."""
    by_id = {task.id: task for task in tasks}
    state: Dict[str, int] = {}
    ordered: List[Task] = []

    def visit(task: Task, trail: List[str]) -> None:
        """Depth-first walk that records a task after its dependencies."""
        mark = state.get(task.id, 0)
        if mark == 2:
            return
        if mark == 1:
            raise CycleError(" -> ".join(trail + [task.id]))
        state[task.id] = 1
        for dep in task.depends_on:
            if dep in by_id:
                visit(by_id[dep], trail + [task.id])
        state[task.id] = 2
        ordered.append(task)

    for task in tasks:
        visit(task, [])
    return ordered


def _dirs_overlap(left: Task, right: Task) -> bool:
    """Whether two tasks claim the same directory, or one claims a parent of the other's."""
    for a in left.dirs:
        for b in right.dirs:
            if a == b or a.startswith(b + "/") or b.startswith(a + "/"):
                return True
    return False


def waves(tasks: Sequence[Task], done: Sequence[str] = ()) -> List[List[Task]]:
    """Groups tasks into waves whose members share no directory and whose dependencies are all satisfied earlier."""
    ordered = [task for task in topological(tasks) if task.id not in done]
    finished = set(done)
    result: List[List[Task]] = []
    pending = list(ordered)
    while pending:
        wave: List[Task] = []
        remaining: List[Task] = []
        for task in pending:
            ready = all(dep in finished or dep not in {t.id for t in tasks} for dep in task.depends_on)
            if ready and not any(_dirs_overlap(task, member) for member in wave):
                wave.append(task)
            else:
                remaining.append(task)
        if not wave:
            raise CycleError("no runnable task among: " + ", ".join(task.id for task in remaining))
        result.append(wave)
        finished.update(task.id for task in wave)
        pending = remaining
    return result


def blocked_by(tasks: Sequence[Task], failed: str) -> List[str]:
    """Ids of every task that transitively depends on the failed one."""
    dependents: Dict[str, List[str]] = {}
    for task in tasks:
        for dep in task.depends_on:
            dependents.setdefault(dep, []).append(task.id)
    out: List[str] = []
    frontier = [failed]
    while frontier:
        current = frontier.pop()
        for child in dependents.get(current, []):
            if child not in out:
                out.append(child)
                frontier.append(child)
    return out

File: internal/line/dag.go
// Package line is the conveyor: intake, the input and output devices, QC, and ship.
package line

import (
	"fmt"
	"strings"

	"komodo/internal/backlog"
)

// Topological orders tasks so every dependency precedes its dependents, file order breaking ties.
func Topological(tasks []backlog.Task) ([]backlog.Task, error) {
	byID := map[string]backlog.Task{}
	for _, task := range tasks {
		byID[task.ID] = task
	}
	state := map[string]int{}
	var ordered []backlog.Task
	var visit func(backlog.Task, []string) error
	visit = func(task backlog.Task, trail []string) error {
		switch state[task.ID] {
		case 2:
			return nil
		case 1:
			return fmt.Errorf("dependency cycle: %s", strings.Join(append(trail, task.ID), " -> "))
		}
		state[task.ID] = 1
		for _, dep := range task.DependsOn() {
			if parent, ok := byID[dep]; ok {
				if err := visit(parent, append(trail, task.ID)); err != nil {
					return err
				}
			}
		}
		state[task.ID] = 2
		ordered = append(ordered, task)
		return nil
	}
	for _, task := range tasks {
		if err := visit(task, nil); err != nil {
			return nil, err
		}
	}
	return ordered, nil
}

// dirsOverlap reports whether two tasks claim the same directory, or one claims a parent of the other's.
func dirsOverlap(left, right backlog.Task) bool {
	for _, a := range left.Dirs() {
		for _, b := range right.Dirs() {
			if a == b || strings.HasPrefix(a, b+"/") || strings.HasPrefix(b, a+"/") {
				return true
			}
		}
	}
	return false
}

// Waves groups tasks so members of one wave share no directory and no unmet dependency.
func Waves(tasks []backlog.Task, done []string) ([][]backlog.Task, error) {
	ordered, err := Topological(tasks)
	if err != nil {
		return nil, err
	}
	known := map[string]bool{}
	for _, task := range tasks {
		known[task.ID] = true
	}
	finished := map[string]bool{}
	for _, id := range done {
		finished[id] = true
	}
	var pending []backlog.Task
	for _, task := range ordered {
		if !finished[task.ID] {
			pending = append(pending, task)
		}
	}
	var result [][]backlog.Task
	for len(pending) > 0 {
		var wave, remaining []backlog.Task
		for _, task := range pending {
			ready := true
			for _, dep := range task.DependsOn() {
				if known[dep] && !finished[dep] {
					ready = false
					break
				}
			}
			clash := false
			for _, member := range wave {
				if dirsOverlap(task, member) {
					clash = true
					break
				}
			}
			if ready && !clash {
				wave = append(wave, task)
				continue
			}
			remaining = append(remaining, task)
		}
		if len(wave) == 0 {
			var ids []string
			for _, task := range remaining {
				ids = append(ids, task.ID)
			}
			return nil, fmt.Errorf("no runnable task among: %s", strings.Join(ids, ", "))
		}
		for _, task := range wave {
			finished[task.ID] = true
		}
		result = append(result, wave)
		pending = remaining
	}
	return result, nil
}

// BlockedBy lists every task that transitively depends on the failed one.
func BlockedBy(tasks []backlog.Task, failed string) []string {
	dependents := map[string][]string{}
	for _, task := range tasks {
		for _, dep := range task.DependsOn() {
			dependents[dep] = append(dependents[dep], task.ID)
		}
	}
	seen := map[string]bool{}
	var out []string
	frontier := []string{failed}
	for len(frontier) > 0 {
		current := frontier[len(frontier)-1]
		frontier = frontier[:len(frontier)-1]
		for _, child := range dependents[current] {
			if !seen[child] {
				seen[child] = true
				out = append(out, child)
				frontier = append(frontier, child)
			}
		}
	}
	return out
}
